### PR TITLE
add option to (re-)use existing app, introduce tests

### PR DESCRIPTION
**Pull request by risicle**
https://www.pivotaltracker.com/story/show/181280979

Originally this story was about giving `conduit` the ability to use an existing app, but it expanded to make the code a lot more amenable to writing tests - which in turn forced me to nail down a lot of subtle details about its behaviour.

Individual commits have more detail, but in summary, a `App.PrepareForExistingApp()` method was added, analogous to `App.DeployApp()`, and responsible for getting the `App` to a comparable state of initialization before the common `App.SetupTunnels()` is called. Much common initialization is moved to a unified `App.initServiceBindings()` method, now called from `App.SetupTunnels()`. This makes it much easier to be certain the logic around binding information checking, program name checking and service name checking is consistent.

Expose this to the user via a new `--existing-app` flag, reusing the existing `--app-name` flag for specifying the app name. The addition of this feature now makes the existence of the `--reuse` flag confusing, as it has only ever controlled whether the app is deleted on teardown - so rename it to `--no-delete`. Fortunately it was previously a hidden flag so it's less likely it's in heavy use. Still, it is supported as a (deprecated) alias for the option.

To make the above testable, some amount of refactoring of the `client` package was required, to use `client.Client` through an interface, making it fake-able, which we do through the added `counterfeiter` dependency. `ifacemaker` is used to generate the interface for `client.Client`. Previous attempts at exposing credentials through an interface were abandoned and `client.Credentials` was made into a public type definition. This is because actually _using_ credentials via an interface was going to cause all sorts of json unmarshalling problems which aren't worth the pain.

Adding the tests required me to settle a lot of the undefined corner case behaviours shown by conduit around e.g. multiple service connections, mismatches of program type etc. In these cases I've tried to choose a behaviour which is both moderately sensible yet also easy to implement.

Despite the addition of (some) tests, this still requires manual testing because there remains plenty of functionality not covered by tests.

### CODE REVIEW

**Comment by cadmiumcat** [comment]
Any chance you'd remove the commits with interface stuff that we don't need to care about? it might make it easier for review. Unless there's a reason for keeping them

**Comment by risicle** [comment]
The interface stuff is used by the tests.

In the "Files changed" tab you can hit the "Changes from all commits" dropdown to shift-click a subset of commits to view at once.

**Comment by dark5un** [comment]
👍

File: README.md [hunk 1]
```diff
@@ -66,6 +66,13 @@ cf conduit service-1 service-2
 
 Output from the command will report connection details for the tunnel(s) in the foreground, hit Ctrl+C to terminate the connections.
 
+### Conduit apps
+
+Traditionally `cf-conduit` creates an app to implement the tunnel. This app can be named with the `--app-name` option. This app will be deleted when the tunnel is closed. `--no-delete` option stops `cf-conduit` from deleting the app when the tunnel closes.
+
+### Reusing existing app
+
+An existing app can be reused, providing `--existing-app` flag with `--app-name`.
```
Comment (risicle) [inline README.md:75]: Worth mentioning that this app needs to already be bound to the destination service(s)

Comment (dark5un) [reply README.md:75]: good catch

**Comment by risicle** [comment]
:+1: cool (it won't let me approve as it's my own PR)
